# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webscheduling/WebTaskScheduler.h
# Commit: 0773f669c498
# Full Hash: 0773f669c4983032149421e810a114d573614c07
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2025-02-27 21:52:52
# Regressor Bug: 1920115
# File Overlap Count: 2
# Description:
#   Bug 1920115 - Allow selecting tasks from all existing web schedulers for main thread r=jjaschke,smaug
#   
#   This is a missing step in the current implementation. When we about
#   to select a task, it should pick one from all scheduled tasks, including
#   the ones scheduled by different schedulers.
# ==============================================================================

diff -r 4b579a1c9b46 -r 0773f669c498 dom/webscheduling/WebTaskScheduler.h
--- a/dom/webscheduling/WebTaskScheduler.h	Thu Feb 27 16:17:35 2025 +0000
+++ b/dom/webscheduling/WebTaskScheduler.h	Thu Feb 27 16:17:36 2025 +0000
@@ -246,7 +246,9 @@
 class WebTaskSchedulerMainThread;
 class WebTaskSchedulerWorker;
 
-class WebTaskScheduler : public nsWrapperCache, public SupportsWeakPtr {
+class WebTaskScheduler : public nsWrapperCache,
+                         public SupportsWeakPtr,
+                         public LinkedListElement<WebTaskScheduler> {
   friend class DelayedWebTaskHandler;
 
  public:
@@ -271,7 +273,7 @@
   virtual JSObject* WrapObject(JSContext* cx,
                                JS::Handle<JSObject*> aGivenProto) override;
 
-  WebTask* GetNextTask();
+  WebTask* GetNextTask(bool aIsMainThread);
 
   virtual void Disconnect();
 
@@ -288,8 +290,6 @@
   virtual ~WebTaskScheduler() = default;
   nsCOMPtr<nsIGlobalObject> mParent;
 
-  uint32_t mNextEnqueueOrder;
-
  private:
   struct SelectedTaskQueueData {
     WebTaskQueueHashKey mSelectedQueueHashKey;
@@ -314,6 +314,10 @@
 
   EventQueuePriority GetEventQueuePriority(const TaskPriority& aPriority) const;
 
+  nsTHashMap<WebTaskQueueHashKey, WebTaskQueue>& GetWebTaskQueues() {
+    return mWebTaskQueues;
+  }
+
   nsTHashMap<WebTaskQueueHashKey, WebTaskQueue> mWebTaskQueues;
 };
 